File: leetcode/design/353_Design_Snake_Game/codes/design.cpp
#include <vector>
#include <deque>
#include <unordered_set>
#include <string>
using namespace std;

class SnakeGame {
    int width, height, score = 0, foodIndex = 0;
    vector<vector<int>> food;
    deque<pair<int, int>> snake;
    unordered_set<int> snakeSet;

    int hash(int x, int y) { return x * width + y; }

public:
    SnakeGame(int width, int height, vector<vector<int>>& food)
        : width(width), height(height), food(food) {
        snake.push_front({0, 0});
        snakeSet.insert(hash(0, 0));
    }

    int move(string direction) {
        int dx = 0, dy = 0;
        if (direction == "U") dx = -1;
        else if (direction == "D") dx = 1;
        else if (direction == "L") dy = -1;
        else if (direction == "R") dy = 1;

        int nx = snake.front().first + dx;
        int ny = snake.front().second + dy;
        int newHead = hash(nx, ny);

        if (nx < 0 || ny < 0 || nx >= height || ny >= width ||
            (snakeSet.count(newHead) && make_pair(nx, ny) != snake.back()))
            return -1;

        if (foodIndex < food.size() && food[foodIndex][0] == nx && food[foodIndex][1] == ny) {
            foodIndex++;
            score++;
        } else {
            auto tail = snake.back();
            snakeSet.erase(hash(tail.first, tail.second));
            snake.pop_back();
        }

        snake.push_front({nx, ny});
        snakeSet.insert(newHead);
        return score;
    }
};


/**
 * Your SnakeGame object will be instantiated and called as such:
 * SnakeGame* obj = new SnakeGame(width, height, food);
 * int param_1 = obj -> move(direction);
 */

File: leetcode/design/353_Design_Snake_Game/codes/design.py
from collections import deque

class SnakeGame:

    def __init__(self, width: int, height: int, food: list[list[int]]):
        self.width = width
        self.height = height
        self.food = food
        self.food_index = 0
        self.score = 0

        self.snake = deque([(0, 0)])
        self.snake_set = {(0, 0)}

        self.dirs = {'U': (-1, 0), 'D': (1, 0), 'L': (0, -1), 'R': (0, 1)}

    def move(self, direction: str) -> int:
        dx, dy = self.dirs[direction]
        head_x, head_y = self.snake[0]
        new_head = (head_x + dx, head_y + dy)

        if (not (0 <= new_head[0] < self.height) or
            not (0 <= new_head[1] < self.width) or
            (new_head in self.snake_set and new_head != self.snake[-1])):
            return -1

        if self.food_index < len(self.food) and self.food[self.food_index] == list(new_head):
            self.food_index += 1
            self.score += 1
        else:
            tail = self.snake.pop()
            self.snake_set.remove(tail)

        self.snake.appendleft(new_head)
        self.snake_set.add(new_head)

        return self.score

# Your SnakeGame object will be instantiated and called as such:
# obj = SnakeGame(width, height, food)
# param_1 = obj.move(direction)